allennlp/training/cross_validation.py

```diff
@@ -0,0 +1,317 @@
+        random_state: Optional[Union[int, RandomState]] = None,
+    ) -> None:
+        super().__init__(
+            n_splits=n_splits, test_size=test_size, train_size=train_size, random_state=random_state
+        )
+
+
+@CrossValidator.register("time_series_split")
+class TimeSeriesSplit(CrossValidator, sklearn.model_selection.TimeSeriesSplit):
+    def __init__(self, n_splits: int = 5, max_train_size: Optional[int] = None) -> None:
+        super().__init__(n_splits=n_splits, max_train_size=max_train_size)
+
+
+logger = logging.getLogger(__name__)
+
+
+@TrainModel.register("cross_validation", constructor="from_partial_objects")
+class CrossValidateModel(TrainModel):
+    def __init__(
+        self,
+        serialization_dir: str,
+        dataset: Dataset,
+        data_loader_builder: Lazy[DataLoader],
+        model: Model,
+        trainer_builder: Lazy[Trainer],
+        cross_validator: CrossValidator,
+        batch_weight_key: str = "",
+        retrain: bool = False,
+    ) -> None:
+        data_loader = data_loader_builder.construct(dataset=dataset)
+
+        # The "main" trainer. It's used to pass to super and when `retrain` option is set.
+        #
+        # We don't need to pass `serialization_dir` and `local_rank` here, because they will have
+        # been passed through the trainer by from_params already, because they were keyword
+        # arguments to construct this class in the first place.
+        trainer = trainer_builder.construct(model=model, data_loader=data_loader)
+
+        super().__init__(serialization_dir=serialization_dir, model=model, trainer=trainer)
+
+        self.dataset = dataset
+        self.data_loader_builder = data_loader_builder
+        self.trainer_builder = trainer_builder
+        self.batch_weight_key = batch_weight_key
+        self.cross_validator = cross_validator
+        self.retrain = retrain
+
+    @overrides
+    def run(self) -> Dict[str, Any]:
+        metrics_by_fold = []
+
+        n_splits = self.cross_validator.get_n_splits(self.dataset)
+
+        for fold_index, (train_indices, test_indices) in enumerate(
+            self.cross_validator(self.dataset)
+        ):
+            logger.info(f"Fold {fold_index}/{n_splits - 1}")
+
+            serialization_dir = os.path.join(self.serialization_dir, f"fold_{fold_index}")
+            os.makedirs(serialization_dir, exist_ok=True)
+
+            train_dataset = Subset(self.dataset, train_indices)
+            test_dataset = Subset(self.dataset, test_indices)
+
+            train_data_loader = self.data_loader_builder.construct(dataset=train_dataset)
+            test_data_loader = self.data_loader_builder.construct(dataset=test_dataset)
+
+            model = copy.deepcopy(self.model)
+            subtrainer = self.trainer_builder.construct(
+                serialization_dir=serialization_dir, data_loader=train_data_loader, model=model
+            )
+
+            fold_metrics = subtrainer.train()
```

Comment:
We could add support for nested cross-validation, which should be as easy as supporting another `CrossValidator` and using it.

Reply:
Let's get the 90% case working first.